File: ruby/medium/binary_tree_inorder_traversal.rb

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}
def inorder_traversal(root)
    stack = Array.new
    res = Array.new
    go_left(stack, root)
    while stack.length > 0
        top = stack.pop
        res.push(top.val)
        go_left(stack, top.right)
    end
    return res
end

def go_left(stack, node)
    while(node)
        stack.push(node)
        node = node.left
    end
end

File: javascript/medium/binaryTreeInorderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
    const stack = new Array(), res = new Array();
    goLeft(stack, root);
    while(stack.length > 0) {
        let top = stack.pop();
        res.push(top.val);
        goLeft(stack, top.right);
    }
    return res;
};

var goLeft = function(stack, root) {
    while(root !== null) {
        stack.push(root);
        root = root.left;
    }
};
